pkg/helpers/crd.go

```diff
@@ -102,3 +124,89 @@ func (s SecretSortList) Less(i, j int) bool {
 
 	return s.Items[i].CreationTimestamp.Before(&s.Items[j].CreationTimestamp)
 }
+
+// Checks if the slice of strings contains a specific string
+func SliceContainsString(list []string, s string) bool {
+	for _, v := range list {
+		if v == s {
+			return true
+		}
+	}
+	return false
+}
+
+// handleDeletion will handle the finalizers if the object was deleted.
+// It will return true, in case the object has been deleted and the finalizers
+// were removed. If the object was removed the reconcile can be returned.
+func HandleDeletion(instance metav1.Object, resource Remover, finalizerName string, client client.Client) (bool, error) {
+	if instance.GetDeletionTimestamp() == nil {
+		return false, nil
+	}
+
+	annotationValue, exists := instance.GetAnnotations()[DeleteProtectionAnnotation]
+
+	// this is really weird: https://play.golang.org/p/NJkbjN7fCMa
+	var protected bool
+	var err error
+	if exists {
+		protected, err = strconv.ParseBool(annotationValue)
+		// Assume true if it can't be parsed
+		if err != nil {
+			protected = true
+			// We need to reset the error again, so we don't trigger any unwanted side effects...
+			err = nil
+		}
+	} else {
+		protected = false
+	}
+
+	deleted := false
+	if SliceContainsString(instance.GetFinalizers(), finalizerName) {
+		if !protected {
```

Comment:
If it *is* protected, we shouldn't remove the finalizer, since this would delete the object. So this if should probably include the whole function.

Reply:
Yes, but that's fine in my opinion. It's just how finalizers work. Once we have validating webhooks, we can improve the UX by not deleting the object at all as long as the annotation exists.

Reply:
If it won't cause the API to fall over then it's also fine by me. I'll change it tomorrow.